File: src/personal_finance/theme.py
"""Refined Swiss Banking aesthetic for the dashboard.

Design direction: "Midnight Vault" - Dark mode variant
- Deep charcoal/obsidian backgrounds with warm undertones
- Burnished gold/champagne accents for luxurious warmth
- Bold, oversized numbers with serif elegance
- Generous negative space and precise geometry
- Subtle ambient glow for private banking atmosphere
"""

from decimal import Decimal
from typing import Union

# Type alias for numeric values that can be Decimal or float
NumericValue = Union[Decimal, float]

# Google Fonts import URL - Adding Playfair Display for refined dark mode headers
FONTS_URL = "https://fonts.googleapis.com/css2?family=Cormorant+Garamond:wght@400;600;700&family=DM+Sans:wght@400;500;600&family=Playfair+Display:wght@400;500;600&display=swap"

# Color palette - Midnight Vault (Dark Mode)
COLORS = {
    # Backgrounds - Deep obsidian with warm undertones
    "background": "#0D0D0F",  # Near-black with slight warmth
    "card": "#18181B",  # Elevated dark surface (zinc-900)
    "card_elevated": "#1F1F23",  # Hover/elevated state
    # Text - High contrast with warm whites
    "text_primary": "#FAFAF9",  # Warm white (stone-50)
    "text_secondary": "#A8A29E",  # Warm gray (stone-400)
    "text_muted": "#78716C",  # Dimmed warm gray (stone-500)
    # Accents - Burnished gold/champagne
    "accent": "#D4A853",  # Burnished gold
    "accent_light": "#E5C06E",  # Lighter champagne
    "accent_glow": "rgba(212, 168, 83, 0.15)",  # Gold ambient glow
    # Semantic - Muted tones for refined dark aesthetic
    "positive": "#6EBF8B",  # Sage green (sophisticated, less saturated)
    "positive_bg": "rgba(110, 191, 139, 0.12)",
    "negative": "#E07A7A",  # Dusty rose (softer coral)
    "negative_bg": "rgba(224, 122, 122, 0.12)",
    # Charts - Muted, sophisticated palette for professional finance
    "chart_1": "#D4A853",  # Burnished gold (primary accent - unchanged)
    "chart_2": "#7BA3C9",  # Slate blue (desaturated sapphire)
    "chart_3": "#6EBF8B",  # Sage green (muted emerald)
    "chart_4": "#D4956A",  # Terracotta (muted amber)
    # Borders and lines - Subtle dark borders
    "border": "#27272A",  # Zinc-800
    "border_strong": "#3F3F46",  # Zinc-700
    "divider": "#1F1F23",  # Subtle divider
}

# Typography - Enhanced for dark mode legibility
FONTS = {
    "display": "'Playfair Display', 'Cormorant Garamond', Georgia, serif",
    "body": "'DM Sans', -apple-system, BlinkMacSystemFont, sans-serif",
}

# Plotly chart template - Dark and refined
CHART_TEMPLATE = {
    "layout": {
        "paper_bgcolor": COLORS["card"],
        "plot_bgcolor": COLORS["card"],
        "font": {
            "family": FONTS["body"],
            "color": COLORS["text_secondary"],
            "size": 12,
        },
        "title": {
            "font": {
                "family": FONTS["display"],
                "size": 20,
                "color": COLORS["text_primary"],
            },
            "x": 0,
            "xanchor": "left",
        },
        "xaxis": {
            "gridcolor": COLORS["divider"],
            "linecolor": COLORS["border"],
            "tickfont": {"size": 11, "color": COLORS["text_muted"]},
            "showgrid": False,
            "zeroline": False,
        },
        "yaxis": {
            "gridcolor": COLORS["border"],
            "linecolor": "rgba(0,0,0,0)",
            "tickfont": {"size": 11, "color": COLORS["text_muted"]},
            "showgrid": True,
            "gridwidth": 1,
            "zeroline": False,
        },
        "legend": {
            "bgcolor": "rgba(0,0,0,0)",
            "font": {"color": COLORS["text_secondary"], "size": 11},
            "orientation": "h",
            "yanchor": "bottom",
            "y": 1.02,
            "xanchor": "left",
            "x": 0,
        },
        "margin": {"t": 60, "r": 24, "b": 48, "l": 60},
        "hoverlabel": {
            "bgcolor": COLORS["card_elevated"],
            "font": {"color": COLORS["text_primary"], "family": FONTS["body"]},
            "bordercolor": COLORS["border_strong"],
        },
    }
}

# CSS styles - Dark mode refinement
STYLES = {
    "page": {
        "backgroundColor": COLORS["background"],
        "minHeight": "100vh",
        "padding": "40px 48px",
        "fontFamily": FONTS["body"],
        "color": COLORS["text_primary"],
        # Subtle radial gradient for depth
        "backgroundImage": "radial-gradient(ellipse at 50% 0%, rgba(212, 168, 83, 0.03) 0%, transparent 50%)",
    },
    "card": {
        "backgroundColor": COLORS["card"],
        "borderRadius": "3px",
        "padding": "32px",
        "marginBottom": "20px",
        "border": f"1px solid {COLORS['border']}",
        "boxShadow": "0 4px 24px rgba(0, 0, 0, 0.4), 0 1px 2px rgba(0, 0, 0, 0.3)",
        "position": "relative",
        # Subtle inner glow on top edge
        "backgroundImage": f"linear-gradient(180deg, {COLORS['card_elevated']} 0%, {COLORS['card']} 2px)",
    },
    "metric_value": {
        "fontSize": "42px",
        "fontWeight": "500",
        "fontFamily": FONTS["display"],
        "color": COLORS["text_primary"],
        "margin": "0",
        "letterSpacing": "-0.02em",
        "lineHeight": "1.1",
        # Subtle text shadow for depth
        "textShadow": "0 2px 4px rgba(0, 0, 0, 0.3)",
    },
    "metric_label": {
        "fontSize": "11px",
        "fontWeight": "600",
        "fontFamily": FONTS["body"],
        "color": COLORS["text_muted"],
        "marginBottom": "12px",
        "textTransform": "uppercase",
        "letterSpacing": "0.12em",
    },
    "metric_change_positive": {
        "fontSize": "13px",
        "fontWeight": "500",
        "color": COLORS["positive"],
        "marginTop": "12px",
        "padding": "6px 12px",
        "backgroundColor": COLORS["positive_bg"],
        "borderRadius": "3px",
        "display": "inline-block",
        "border": f"1px solid rgba(110, 191, 139, 0.2)",
    },
    "metric_change_negative": {
        "fontSize": "13px",
        "fontWeight": "500",
        "color": COLORS["negative"],
        "marginTop": "12px",
        "padding": "6px 12px",
        "backgroundColor": COLORS["negative_bg"],
        "borderRadius": "3px",
        "display": "inline-block",
        "border": f"1px solid rgba(224, 122, 122, 0.2)",
    },
    "tab": {
        "backgroundColor": "transparent",
        "color": COLORS["text_muted"],
        "border": "none",
        "borderBottom": "2px solid transparent",
        "borderRadius": "0",
        "padding": "16px 24px",
        "fontFamily": FONTS["body"],
        "fontSize": "13px",
        "fontWeight": "500",
        "textTransform": "uppercase",
        "letterSpacing": "0.1em",
        "transition": "all 0.25s ease",
    },
    "tab_selected": {
        "backgroundColor": "transparent",
        "color": COLORS["accent"],
        "border": "none",
        "borderBottom": f"2px solid {COLORS['accent']}",
        "borderRadius": "0",
        "padding": "16px 24px",
        "fontFamily": FONTS["body"],
        "fontSize": "13px",
        "fontWeight": "600",
        "textTransform": "uppercase",
        "letterSpacing": "0.1em",
        # Subtle glow effect on selected tab
        "textShadow": f"0 0 20px {COLORS['accent_glow']}",
    },
    "header": {
        "display": "flex",
        "justifyContent": "space-between",
        "alignItems": "center",
        "marginBottom": "48px",
        "paddingBottom": "24px",
        "borderBottom": f"1px solid {COLORS['border']}",
    },
    "title": {
        "fontSize": "32px",
        "fontWeight": "500",
        "fontFamily": FONTS["display"],
        "color": COLORS["text_primary"],
        "margin": "0",
        "letterSpacing": "-0.01em",
        # Subtle gold accent on title
        "backgroundImage": f"linear-gradient(135deg, {COLORS['text_primary']} 0%, {COLORS['accent_light']} 100%)",
        "WebkitBackgroundClip": "text",
        "WebkitTextFillColor": "transparent",
        "backgroundClip": "text",
    },
    "grid": {
        "display": "grid",
        "gridTemplateColumns": "repeat(3, 1fr)",

        # "gridTemplateColumns": "repeat(auto-fit, minmax(280px, 1fr))",
        "gap": "20px",
        "marginBottom": "32px",
    },
    "chart_container": {
        "backgroundColor": COLORS["card"],
        "borderRadius": "3px",
        "padding": "24px",
        "marginBottom": "20px",
        "border": f"1px solid {COLORS['border']}",
        "boxShadow": "0 4px 24px rgba(0, 0, 0, 0.4), 0 1px 2px rgba(0, 0, 0, 0.3)",
        # Subtle inner glow
        "backgroundImage": f"linear-gradient(180deg, {COLORS['card_elevated']} 0%, {COLORS['card']} 2px)",
    },
    "upload_button": {
        "padding": "12px 24px",
        "backgroundColor": COLORS["card"],
        "border": f"1px solid {COLORS['border_strong']}",
        "borderRadius": "3px",
        "cursor": "pointer",
        "color": COLORS["text_secondary"],
        "fontFamily": FONTS["body"],
        "fontSize": "13px",
        "fontWeight": "500",
        "letterSpacing": "0.02em",
        "transition": "all 0.25s ease",
        "boxShadow": "0 2px 8px rgba(0, 0, 0, 0.3)",
    },
    "empty_state": {
        "textAlign": "center",
        "padding": "80px 40px",
        "backgroundColor": COLORS["card"],
        "border": f"1px solid {COLORS['border']}",
        "borderRadius": "3px",
        "boxShadow": "0 4px 24px rgba(0, 0, 0, 0.4)",
    },
    "tabs_container": {
        "borderBottom": f"1px solid {COLORS['border']}",
        "marginBottom": "32px",
    },
}


def format_currency(value: NumericValue) -> str:
    """Format a number as USD currency.

    Accepts both Decimal and float values for compatibility.
    """
    # Convert to float for formatting (Decimal formatting would be identical)
    v = float(value)
    if abs(v) >= 1_000_000:
        return f"${v / 1_000_000:,.2f}M"
    elif abs(v) >= 1_000:
        return f"${v / 1_000:,.1f}K"
    else:
        return f"${v:,.0f}"


def format_percentage(value: NumericValue) -> str:
    """Format a number as a percentage.

    Accepts both Decimal and float values for compatibility.
    """
    return f"{float(value):+.1f}%"


def format_change(value: NumericValue, is_percentage: bool = False, invert_colors: bool = False) -> tuple[str, dict]:
    """Format a change value with appropriate styling.

    Args:
        value: The change value (Decimal or float)
        is_percentage: If True, format as percentage
        invert_colors: If True, positive values are red (bad) and negative are green (good)
    """
    if is_percentage:
        text = format_percentage(value)
    else:
        sign = "+" if value >= 0 else ""
        text = f"{sign}{format_currency(value)}"

    is_positive = value >= 0
    if invert_colors:
        is_positive = not is_positive

    style = STYLES["metric_change_positive"] if is_positive else STYLES["metric_change_negative"]
    return text, style

File: static-site/src/lib/theme.ts
/**
 * Theme system ported from Python theme.py
 * "Midnight Vault" - Swiss Banking aesthetic in dark mode
 * "Ivory Ledger" - Warm parchment aesthetic in light mode
 */

import Decimal from 'decimal.js';

// Color palette - Midnight Vault (Dark Mode)
export const COLORS = {
  // Backgrounds - Deep obsidian with warm undertones
  background: '#0D0D0F',
  card: '#18181B',
  cardElevated: '#1F1F23',

  // Text - High contrast with warm whites
  textPrimary: '#FAFAF9',
  textSecondary: '#A8A29E',
  textMuted: '#78716C',

  // Accents - Burnished gold/champagne
  accent: '#D4A853',
  accentLight: '#E5C06E',
  accentGlow: 'rgba(212, 168, 83, 0.15)',

  // Semantic - Muted tones for refined dark aesthetic
  positive: '#6EBF8B',
  positiveBg: 'rgba(110, 191, 139, 0.12)',
  negative: '#E07A7A',
  negativeBg: 'rgba(224, 122, 122, 0.12)',

  // Charts - Muted, sophisticated palette for professional finance
  chart1: '#D4A853',
  chart2: '#7BA3C9',
  chart3: '#6EBF8B',
  chart4: '#D4956A',
  chart5: '#B8A9C9', // muted lavender
  chart6: '#D4956A', // terracotta
  chart7: '#7BA3A3', // muted teal
  chart8: '#8B9EB3', // cool slate
  chart9: '#C9A86C', // muted ochre

  // Borders and lines - Subtle dark borders
  border: '#27272A',
  borderStrong: '#3F3F46',
  divider: '#1F1F23',
} as const;

// Color palette - Ivory Ledger (Light Mode)
export const LIGHT_COLORS = {
  // Backgrounds - Warm parchment
  background: '#F5F2EB',
  card: '#FAFAF5',
  cardElevated: '#FFFFFF',

  // Text - Dark warm tones for readability
  textPrimary: '#1A1714',
  textSecondary: '#6B6560',
  textMuted: '#9B918A',

  // Accents - Deeper gold for light backgrounds
  accent: '#B07A1A',
  accentLight: '#C99027',
  accentGlow: 'rgba(176, 122, 26, 0.10)',

  // Semantic - Richer tones for light aesthetic
  positive: '#1E7A45',
  positiveBg: 'rgba(30, 122, 69, 0.08)',
  negative: '#B83232',
  negativeBg: 'rgba(184, 50, 50, 0.08)',

  // Charts - Deeper saturated palette for light backgrounds
  chart1: '#B07A1A',
  chart2: '#3A6B96',
  chart3: '#1E7A45',
  chart4: '#B8622A',
  chart5: '#7A5B9B',
  chart6: '#B8622A',
  chart7: '#3D8A8A',
  chart8: '#516080',
  chart9: '#9E7520',

  // Borders - Warm light borders
  border: '#DDD8CE',
  borderStrong: '#C5BEB4',
  divider: '#EAE5DC',
} as const;

// Typography
export const FONTS = {
  display: "'Playfair Display', 'Cormorant Garamond', Georgia, serif",
  body: "'DM Sans', -apple-system, BlinkMacSystemFont, sans-serif",
} as const;

// Plotly chart template - Dark and refined
export const CHART_TEMPLATE = {
  paper_bgcolor: COLORS.card,
  plot_bgcolor: COLORS.card,
  font: {
    family: FONTS.body,
    color: COLORS.textSecondary,
    size: 12,
  },
  title: {
    font: {
      family: FONTS.display,
      size: 20,
      color: COLORS.textPrimary,
    },
    x: 0,
    xanchor: 'left' as const,
  },
  xaxis: {
    gridcolor: COLORS.divider,
    linecolor: COLORS.border,
    tickfont: { size: 14, color: COLORS.textMuted },
    showgrid: false,
    zeroline: false,
  },
  yaxis: {
    gridcolor: COLORS.border,
    linecolor: 'rgba(0,0,0,0)',
    tickfont: { size: 14, color: COLORS.textMuted },
    showgrid: true,
    gridwidth: 1,
    zeroline: false,
  },
  legend: {
    bgcolor: 'rgba(0,0,0,0)',
    font: { color: COLORS.textSecondary, size: 11 },
    orientation: 'h' as const,
    yanchor: 'bottom' as const,
    y: 1.02,
    xanchor: 'left' as const,
    x: 0,
  },
  margin: { t: 60, r: 24, b: 48, l: 60 },
  hoverlabel: {
    bgcolor: COLORS.cardElevated,
    font: { color: COLORS.textPrimary, family: FONTS.body },
    bordercolor: COLORS.borderStrong,
  },
};

// Theme-aware color and template factories
export function getColors(isDark: boolean) {
  return isDark ? COLORS : LIGHT_COLORS;
}

export function getChartTemplate(isDark: boolean) {
  const c = getColors(isDark);
  return {
    paper_bgcolor: c.card,
    plot_bgcolor: c.card,
    font: {
      family: FONTS.body,
      color: c.textSecondary,
      size: 12,
    },
    title: {
      font: {
        family: FONTS.display,
        size: 20,
        color: c.textPrimary,
      },
      x: 0,
      xanchor: 'left' as const,
    },
    xaxis: {
      gridcolor: c.divider,
      linecolor: c.border,
      tickfont: { size: 14, color: c.textMuted },
      showgrid: false,
      zeroline: false,
    },
    yaxis: {
      gridcolor: c.border,
      linecolor: 'rgba(0,0,0,0)',
      tickfont: { size: 14, color: c.textMuted },
      showgrid: true,
      gridwidth: 1,
      zeroline: false,
    },
    legend: {
      bgcolor: 'rgba(0,0,0,0)',
      font: { color: c.textSecondary, size: 11 },
      orientation: 'h' as const,
      yanchor: 'bottom' as const,
      y: 1.02,
      xanchor: 'left' as const,
      x: 0,
    },
    margin: { t: 60, r: 24, b: 48, l: 60 },
    hoverlabel: {
      bgcolor: c.cardElevated,
      font: { color: c.textPrimary, family: FONTS.body },
      bordercolor: c.borderStrong,
    },
  };
}

// FIRE Goal - hardcoded as in Python version
export const FIRE_GOAL = new Decimal('4250000');

// SWR rates for sensitivity analysis
export const SWR_RATES = [
  new Decimal('0.03'),
  new Decimal('0.035'),
  new Decimal('0.04'),
  new Decimal('0.045'),
];

/**
 * Format a number as USD currency.
 * Matches Python theme.py format_currency()
 */
export function formatCurrency(value: Decimal | number): string {
  const v = typeof value === 'number' ? value : value.toNumber();
  if (Math.abs(v) >= 1_000_000) {
    return `$${(v / 1_000_000).toFixed(2)}M`;
  } else if (Math.abs(v) >= 1_000) {
    return `$${(v / 1_000).toFixed(1)}K`;
  } else {
    return `$${v.toLocaleString('en-US', { minimumFractionDigits: 0, maximumFractionDigits: 0 })}`;
  }
}

/**
 * Format a number as a percentage.
 * Matches Python theme.py format_percentage()
 */
export function formatPercentage(value: Decimal | number): string {
  const v = typeof value === 'number' ? value : value.toNumber();
  return `${v >= 0 ? '+' : ''}${v.toFixed(1)}%`;
}

export interface ChangeDisplay {
  text: string;
  isPositive: boolean;
}

/**
 * Format a change value with appropriate styling info.
 * Matches Python theme.py format_change()
 */
export function formatChange(
  value: Decimal | number,
  isPercentage: boolean = false,
  invertColors: boolean = false,
  absoluteValue?: Decimal | number
): ChangeDisplay {
  const v = typeof value === 'number' ? value : value.toNumber();

  let text: string;
  if (isPercentage) {
    text = formatPercentage(v);
  } else {
    const sign = v >= 0 ? '+' : '';
    text = `${sign}${formatCurrency(v)}`;
  }

  // Combine absolute and percentage if both provided
  if (absoluteValue !== undefined) {
    const absV = typeof absoluteValue === 'number' ? absoluteValue : absoluteValue.toNumber();
    const absSign = absV >= 0 ? '+' : '';
    text = `${absSign}${formatCurrency(absV)} / ${formatPercentage(v)}`;
  }

  let isPositive = v >= 0;
  if (invertColors) {
    isPositive = !isPositive;
  }

  return { text, isPositive };
}
